software/spmd/pc_profiler/profiler.S

```diff
@@ -0,0 +1,232 @@
+/* 
+  Description:
+  A software-based PC profiler which builds a histogram of PCs encountered by the program
+
+  Memory layout:
+  DRAM --> 16MB (configurable) for maintaing a histogram for N tiles. 
+
+  Registers used:
+  The program being profiled is free to use any registers within if there are no other imposed constraints.
+  The profiler uses the following registers - 
+  t0 --> Base address of the histogram array (actual value stored in dmem[1])
+  t1 --> Scratchpad (actual value stored in dmem[0])
+*/
+
+#include "bsg_manycore_arch.h"
+#include "bsg_manycore_asm.h"
+
+// Useful to control the amount of space required to simulate the use of the profiler.
+// We want 16MB when running on actual hardware.
+#ifndef HIST_SPACE
+#define HIST_SPACE 16777216
+#endif
+
+#ifndef PROFILE
+#define PROFILE 0
+#endif
+
+#if PROFILE > 0
+/* 
+  Data section:
+  Defines a 2 element (8 byte) array to store the base pointer of the histogram array and for use by the interrupt handler to spill some registers
+  dmem[0] = histogram base pointer, temporary storage
+  dmem[1] = temporary storage
+*/
+.section .dmem.interrupt, "aw"
+  _interrupt_arr: .space 8, 0x0
+
+/*
+  DRAM section:
+  Defines a 16MB (array size is configurable) chunk in DRAM to store the histogram.
+
+  Lingering question(s):
+  1. Assigned space per PC (i.e bin capacity) as 32-bits. Is this enough, too much or too less?
+  2. Base pointer in DRAM (for now it is mapped to start of DRAM) [Consider static assignment in linker]
+*/
+.section .dram, "aw"
+.globl _histogram_arr
+  _histogram_arr: .space HIST_SPACE, 0x0
+#endif
+
+/*
+  Startup:
+  This is very similar to the code in software/spmd/common/crt.S and adds only the interrupt service routine (ISR) and sets up the DMEM with the histogram base address. 
+*/
+.section .crtbegin
+.globl _start
+
+#if PROFILE > 0
+/*
+  Remote interrupt handler (should map to address 0)
+*/
+_remote_interrupt:
+  j _remote_interrupt_handler
+
+/*
+  Trace interrupt handler (should map to address 4)
+  Clears the pending trace interrupt bit in the MIP CSR and atomically increments the histogram bin count based on the MEPC CSR value
+
+  The bin index in the histogram can be found using the following equation index = base ptr + EPC
+  (Although EPC is 4-byte aligned, we are assigning 4 bytes per bin, so we don't need to do EPC >> 2)
+
+  Lingering concerns:
+  Since we are using the MEPC, we don't have the PC of the first instruction in the program. 
+  (Probably does not matter because it's just 1 datapoint)
+*/
+_trace_interrupt:
+  // Save a register to dmem[1]
+  sw t0, 4(zero)
+
+  // clear mip.trace
+  li t0, 0x20000
+  csrrc zero, mip, t0
```

Comment:
@tommydcjung suggest putting mip.trace at bit 0 of mip register, which I believe is unused, so we can just use a csrrci instruction to clear this; this will speed up the interrupt routine and reduce the cache foot print.

Reply:
*puts on risc-v hat* that’s the user-mode software interrupt bit, would be a non-compliant. Probably fine since you’re not going to run OpenSBI or anything, but should note it somewhere.

Reply:
Agreeing with Dan from the RISC-V compatibility standpoint.

Reply:
To clarify, I'm on the side of changing it and adding to a list of risc-v incompatibilities somewhere.  But it's a hardware change, so separate PR

Reply:
Yeah, I visited this a while back from my consulting days. Basically, there are really two sweet spots in compatibility that matter: full toolchain compatibility and full OS-level compatibility. We are the former. Partial OS-level compatibility does leaves everybody complaining about what you are missing.

Reply:
Agreed, but I would phrase it a little differently. 

- The manycore is mostly compatible with RISC-V unprivileged mode (LR/SC behaves differently than expected, other than that I think it is). This is fine because the compiler won't insert LR/SC automatically.
- The manycore has a custom privileged ISA. (missing ebreak, ecall, missing most CSRs, repurposed fields in CSRs). This is fine and expected because these aren't Unix capable cores, which is what the vanilla RISC-V privileged spec is designed for.